### PR TITLE
Import file from local drive

### PR DESCRIPTION
**Pull request by blocus**
## Description

This PR is about the discussion [15972](https://github.com/directus/directus/discussions/15972#discussioncomment-3864789) where I suggested scanning the upload folder and adding the new files to the database. 

## Changes:
- A new button in the files view asks the API to scan the upload folder. 
sync files button
- A new file service method `sync` to check if the file exists on the database and add it
- A new endpoint `/files/scan` which runs the sync method of the file service to all files in the upload folder

## Type of Change

- [ ] Bugfix
- [ ] Improvement
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [ ] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated. PR: [179](https://github.com/directus/docs/pull/179)

### CODE REVIEW

File: api/src/controllers/files.ts [hunk 1]
```diff
@@ -117,6 +118,20 @@ export const multipartHandler: RequestHandler = (req, res, next) => {
 	}
 };
 
+router.get(
+	'/scan',
+	asyncHandler(async (req, res, next) => {
+		const storageFolder = storage.disk('local');
```
Comment (rijkvanzanten) [inline api/src/controllers/files.ts:124]: We can't rely on a hardcoded storage adapter named `local`. `local` is just the default name, but seeing its user created, we can't rely on that name alone

File: api/src/services/files.ts [hunk 2]
```diff
@@ -124,6 +125,50 @@ export class FilesService extends ItemsService {
...
+				const { height, width, description, title, tags, metadata } = await this.getMetadata(buffer.content);
+
+				payload.height ??= height;
+				payload.width ??= width;
+				payload.description ??= description;
+				payload.title ??= title;
+				payload.tags ??= tags;
+				payload.metadata ??= metadata;
+			}
+
```
Comment (rijkvanzanten) [inline api/src/services/files.ts:161]: Wondering if we can make this bit reusable from the regular upload functionality. All the parsing and saving should be the same.

File: api/src/services/files.ts [hunk 3]
```diff
@@ -124,6 +125,50 @@ export class FilesService extends ItemsService {
...
+	 */
+	async sync(relativePath: string): Promise<any> {
+		const res = await this.knex(this.collection).select('*').where({ filename_disk: relativePath }).first();
+
+		if (!res) {
+			const stats = await storage.disk('local').getStat(relativePath);
+			const fileExtension = path.extname(relativePath);
+
+			const payload: Partial<File> & { filename_download: string; storage: string } = {
+				storage: 'local',
```
Comment (rijkvanzanten) [inline api/src/services/files.ts:140]: hardcoded `local`

File: api/src/services/files.ts [hunk 4]
```diff
@@ -124,6 +125,50 @@ export class FilesService extends ItemsService {
 		return primaryKey;
 	}
 
+	/**
+	 * Sync file from filesystem and update it on database
+	 * If the file doesn't exists on database it will create it
+	 */
+	async sync(relativePath: string): Promise<any> {
```
Comment (rijkvanzanten) [inline api/src/services/files.ts:132]: Little worried about the performance downside of doing this one at a time 🤔 Wondering if we should maybe try doing this in batches of `n` to improve import performance

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Great idea @blocus! There's a couple points we need to address to make this production ready for the core codebase:

- Make sure it works for every storage adapter instead of just the one called `local` (and for the case where you have multiple adapters configured)
- Make sure we don't ingest directus-generated thumbnails, as that would cause duplicate records
- Add unit/e2e tests
- Add docs

File: app/src/modules/files/routes/collection.vue [hunk 5]
```diff
@@ -32,6 +32,7 @@
 				<search-input v-model="search" v-model:filter="filter" collection="directus_files" />
 
 				<add-folder :parent="folder" :disabled="createFolderAllowed !== true" />
+				<scan-folder :parent="folder" :disabled="createFolderAllowed !== true" />
```
[the remark this answers is not in the document]
Comment (blocus) [reply app/src/modules/files/routes/collection.vue:35]: I will remove the disabled props until we figure out what permission it should have

**Comment by blocus** [comment]
> * Make sure it works for every storage adapter instead of just the one called `local` (and for the case where you have multiple adapters configured)
> * Make sure we don't ingest directus-generated thumbnails, as that would cause duplicate records
> * Add unit/e2e tests
> * Add docs

I'm not sure I can make the unit/e2e tests 😥 and for docs where should I make it ?

**Comment by erondpowell** [comment]
> and for docs where should I make it ?

Hey there @blocus ! 
I'm Eron, the technical writer. Welcome to Docs Land!
https://github.com/directus/docs

I'll let you get started by "looking and copying" the styles of the other relevant docs.
Once drafted, I'll take a look and provide any revisions feedback, then takeover for the merge.
Feel free to ping me if you get stuck on anything!

> I'm not sure I can make the unit/e2e tests

@licitdev Weren't you doing a ton of unit tests project recently 🤔  Got any tips?

**Comment by licitdev** [comment]
> I'm not sure I can make the unit/e2e tests

You may add unit tests for the file controller and service in the follow files respectively.

https://github.com/directus/directus/blob/dbfedb56fe2efe44187abfd31165faaa79ae7780/api/src/controllers/files.test.ts#L1-L8

https://github.com/directus/directus/blob/dbfedb56fe2efe44187abfd31165faaa79ae7780/api/src/services/files.test.ts#L1-L8

**Comment by rijkvanzanten** [comment]
Heya! Thank you for your contribution. Unfortunately, we can't accept PRs for improvements/features that don't have a prior accepted Discussion and associated Issue (as per the [Contributing Guidelines](https://github.com/directus/directus/blob/main/contributing.md#pull-requests)). This is to ensure that PRs that are opened will actually get the chance to be reviewed, merged, and maintained by the core maintainers. We can reopen and continue this PR review once the requirements have been met. Thank you for your understanding!

**Pull request closed without merging**